In [94]:
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score

In [95]:
df = pd.read_excel('ogrenci.xlsx')
df.head()

Cinsiyet Ebeveyn Egitim Seviyesi Okul Yemekhanesi Ozel Ders  Matematik  \
0    Kadin           Lisans Mezunu              Yok       Yok         72   
1    Kadin         Onlisans Mezunu              Yok       Var         69   
2    Kadin    Yuksek Lisans Mezunu              Yok       Yok         90   
3    Erkek         Onlisans Mezunu              Var       Yok         47   
4    Erkek         Onlisans Mezunu              Yok       Yok         76   

   Okuma  Yazma  
0     72     74  
1     90     88  
2     95     93  
3     57     44  
4     78     75

In [96]:
df.loc[df['Ebeveyn Egitim Seviyesi']=='Lise Mezunu','Ebeveyn Egitim Seviyesi']= 1
df.loc[df['Ebeveyn Egitim Seviyesi']=='Onlisans Mezunu','Ebeveyn Egitim Seviyesi']= 2
df.loc[df['Ebeveyn Egitim Seviyesi']=='Lisans Mezunu','Ebeveyn Egitim Seviyesi']= 3
df.loc[df['Ebeveyn Egitim Seviyesi']=='Yuksek Lisans Mezunu','Ebeveyn Egitim Seviyesi']= 4
df.loc[df['Cinsiyet']=='Kadin','Cinsiyet']= 0
df.loc[df['Cinsiyet']=='Erkek','Cinsiyet']= 1
df.loc[df['Okul Yemekhanesi']=='Yok','Okul Yemekhanesi']= 0
df.loc[df['Okul Yemekhanesi']=='Var','Okul Yemekhanesi']=1
df.loc[df['Ozel Ders']=='Yok','Ozel Ders']=0
df.loc[df['Ozel Ders']=='Var','Ozel Ders']=1


In [97]:
df.head()

Cinsiyet Ebeveyn Egitim Seviyesi Okul Yemekhanesi Ozel Ders  Matematik  \
0        0                       3                0         0         72   
1        0                       2                0         1         69   
2        0                       4                0         0         90   
3        1                       2                1         0         47   
4        1                       2                0         0         76   

   Okuma  Yazma  
0     72     74  
1     90     88  
2     95     93  
3     57     44  
4     78     75

In [98]:
x=df.drop(['Matematik','Okuma','Yazma'],axis=1)
y=df[['Matematik','Okuma','Yazma']]

In [99]:
X_train,X_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [100]:
dt=DecisionTreeRegressor(random_state=42)
dt2=DecisionTreeRegressor(random_state=42)
dt3=DecisionTreeRegressor(random_state=42)
dt_fit_mat=dt.fit(X_train,y_train['Matematik'])
dt_fit_oku=dt2.fit(X_train,y_train['Okuma'])
dt_fit_yaz=dt3.fit(X_train,y_train['Yazma'])


feature_mat=dt_fit_mat.feature_importances_
feature_oku=dt_fit_oku.feature_importances_
feature_yaz=dt_fit_yaz.feature_importances_

feature_names=x.columns
df_content=pd.DataFrame({
    "Özellik":feature_names,
    "Matematik":feature_mat,
    "Okuma":feature_oku,
    "Yazma":feature_yaz
})


In [83]:
df_content

Özellik  Matematik     Okuma     Yazma
0                 Cinsiyet   0.112060  0.269538  0.288433
1  Ebeveyn Egitim Seviyesi   0.168108  0.204350  0.181711
2         Okul Yemekhanesi   0.542155  0.240580  0.228923
3                Ozel Ders   0.177678  0.285532  0.300933

In [84]:
y_pred=dt.predict(X_test)
y_pred2=dt2.predict(X_test)
y_pred3=dt3.predict(X_test)

accuracy_mat=dt.score(X_test,y_test['Matematik'])
accuracy_oku=dt2.score(X_test,y_test['Okuma'])
accuracy_yaz=dt3.score(X_test,y_test['Yazma'])

print('Matematik Doğruluk Oranı:',accuracy_mat)
print('Okuma Doğruluk Oranı:',accuracy_oku)
print('Yazma Doğruluk Oranı:',accuracy_yaz)

Matematik Doğruluk Oranı: 0.1264715085441882
Okuma Doğruluk Oranı: 0.12839884483990427
Yazma Doğruluk Oranı: 0.2336376025422443


In [75]:
df.columns

Index(['Özellik', 'Özellik Katsayısı'], dtype='object')

In [85]:
x=df[['Okuma']]
y=df[['Matematik','Yazma']]
dt4=DecisionTreeRegressor(random_state=42)
dt_fit=dt4.fit(x,y['Matematik'])
dt5=DecisionTreeRegressor(random_state=42)
dt_fit2=dt5.fit(x,y['Yazma'])
dt_feature=dt_fit.feature_importances_
dt_feature2=dt_fit2.feature_importances_
df_okuma=pd.DataFrame({
    "Özellik":x.columns,
    "Matematik":dt_feature,
    "Yazma":dt_feature2
})
df_okuma

Özellik  Matematik  Yazma
0   Okuma        1.0    1.0

Decision Tree modelimin doğruluk oranı düşük çıktı, bu modelden edindiğim bilgilere göre soruları cevaplayacağım

1.Soru: Özel Ders almanın genel başarıya etkisi nedir?

Cevap: Yemekhanenin matematik başarısına etkisini saymazsak en etkili özellik özel ders alma olduğunu görüyoruz özel ders almanın genel başarıya etkisi yüksek

2.Soru: Hangi faktörlerin başarıya etkisi yoktur?

Cevap: Her faktörün az ya da çok etkisi var hiçbir özellik etkisi yoktur denecek kadar etkisiz değil.

3.Soru:  Hangi faktörler başarıyı en çok etkilemektedir?

Cevap: Matematik dersi için en çok etkileyen Yemekhane faktörü, Okuma ve Yazma için en çok etkileyen Özel Ders faktörü genel olarak sıralama yaparsak eğer
Yemekhane>Özel Ders>Cinsiyet>Ebeveyn Eğitim

4.Soru: Okuma Becerisinin yazma ve matematik üzerindeki etkisi nedir

Cevap: Elde ettiğim veriye göre okuma becerisinin yazma ve matematik üzerinde %100 etkisi var

In [86]:
df = pd.read_excel('ogrenci.xlsx')
df.head()

df.loc[df['Ebeveyn Egitim Seviyesi']=='Lise Mezunu','Ebeveyn Egitim Seviyesi']= 1
df.loc[df['Ebeveyn Egitim Seviyesi']=='Onlisans Mezunu','Ebeveyn Egitim Seviyesi']= 2
df.loc[df['Ebeveyn Egitim Seviyesi']=='Lisans Mezunu','Ebeveyn Egitim Seviyesi']= 3
df.loc[df['Ebeveyn Egitim Seviyesi']=='Yuksek Lisans Mezunu','Ebeveyn Egitim Seviyesi']= 4
df.loc[df['Cinsiyet']=='Kadin','Cinsiyet']= 0
df.loc[df['Cinsiyet']=='Erkek','Cinsiyet']= 1
df.loc[df['Okul Yemekhanesi']=='Yok','Okul Yemekhanesi']= 0
df.loc[df['Okul Yemekhanesi']=='Var','Okul Yemekhanesi']=1
df.loc[df['Ozel Ders']=='Yok','Ozel Ders']=0
df.loc[df['Ozel Ders']=='Var','Ozel Ders']=1

In [89]:
x=df.drop(['Matematik','Okuma','Yazma'],axis=1)
y=df[['Matematik','Okuma','Yazma']]
X_train,X_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

lr=LinearRegression()
feature=lr.fit(X_train,y_train['Matematik'])
lr2=LinearRegression()
feature2=lr2.fit(X_train,y_train['Okuma'])
lr3=LinearRegression()
feature3=lr3.fit(X_train,y_train['Yazma'])
y_pred=lr.predict(X_test)
y_pred2=lr2.predict(X_test)
y_pred3=lr3.predict(X_test)
accuracy=lr.score(X_test,y_test['Matematik'])
accuracy2=lr2.score(X_test,y_test['Okuma'])
accuracy3=lr3.score(X_test,y_test['Yazma'])
print("Matematik Doğruluk Oranı:",accuracy)
print("Okuma Doğruluk Oranı:",accuracy2)
print("Yazma Doğruluk Oranı:",accuracy3)

print("Matematik Katsayı:",feature.coef_)
print("Okuma Katsayı:",feature2.coef_)
print("Yazma Katsayı:",feature3.coef_)






Matematik Doğruluk Oranı: 0.16952636337662141
Okuma Doğruluk Oranı: 0.17387475690205545
Yazma Doğruluk Oranı: 0.2625944143015808
Matematik Katsayı: [  4.63936769   2.64247664 -11.72798382   6.28170549]
Okuma Katsayı: [-7.35133369  2.71222488 -7.56485603  7.80769387]
Yazma Katsayı: [-9.31202527  3.70567104 -8.74427514 10.28668489]


In [93]:
# Okuma becerisinin yazma ve matematik üzerindeki etkisi nedir?
x=df[['Okuma']]
y=df[['Matematik','Yazma']]
lr4=LinearRegression()
feature4=lr4.fit(x,y['Matematik'])
lr5=LinearRegression()
feature5=lr5.fit(x,y['Yazma'])
print("Matematik Katsayı:",feature4.coef_)
print("Yazma Katsayı:",feature5.coef_ )

Matematik Katsayı: [0.8491002]
Yazma Katsayı: [0.99353111]


In [91]:
#Özel ders almanın genel başarıya etkisi
df['Genel Basari']= (df['Matematik']+df['Okuma']+df['Yazma'])/3
x=df[['Ozel Ders']]
y=df['Genel Basari']

lr=LinearRegression()
fit=lr.fit(x,y)

print("Katsayı",fit.coef_)

print("Sabit",fit.intercept_)

Katsayı [7.63051915]
Sabit 65.03894080996884


Lineer Regresyon modelimin doğruluk oranı düşük çıktı, bu modelden edindiğim bilgilere göre soruları cevaplayacağım

1.Soru: Özel Ders almanın genel başarıya etkisi nedir?

Cevap: Özel ders almanın genel başarıya etkisi olarak 7.63 katsayı değerini aldığını ve bir hayli etkili olduğunu söyleyebilirim, ilk lineer regresyon sonucuna bakınca en etkili katsayının özel derste olduğunu yine söyleyebilirim

2.Soru: Hangi faktörlerin başarıya etkisi yoktur?

Cevap: Her özelliğin etkisi var ancak bu etkiler çoğunlukla - yönde etki ediyor , etkisiz diyemeyiz ama olumsuz etkiliyor diyebiliriz

3.Soru:  Hangi faktörler başarıyı en çok etkilemektedir?

Cevap: Decision treenin tersine bir sıralama var bu sefer , en çok etkileyen özel ders faktörü en olumsuz etkileyen okul yemekhanesi durumu, sıralama olarak ise
Özel Ders>Ebeveyn Eğitimi>Cinsiyet>Yemekhane

4.Soru: Okuma Becerisinin yazma ve matematik üzerindeki etkisi nedir

Cevap: Elde ettiğim veriye göre okuma becerisinin yazmaya etkisi %99 iken matematiğe etkisi ondan daha az ama yine de bir hayli fazla %85 oranında.
Bu veriye göre okuma başarısı yazma ve matematiği bir hayli etkiliyor

Decision Tree ve Lineer Regresyon modellerini karşılaştıracak olursam eğer
Lineer Regresyondan elde ettiğim verilerin soruları cevaplamamda daha fazla yardımcı olduğunu ve veriyi okumada daha başarılı olduğunu söyleyebilirim.